### PR TITLE
Bug fix: Disease-gene associations

### LINKED ISSUE #156: Gene association representation in `omim.ttl`
I think this is a bug, if not please explain the design decision. For some OMIM disease entries, e.g. https://omim.org/entry/613659, in the omim.ttl file there is only 1 disease to gene association ('has material basis in germline mutation in' IL1B). However, on the OMIM entry page there are 11. I do see all associations in the various data files that are downloaded when creating the omim.ttl file using `python3 -m omim2obo`. 

This issue is not limited to OMIM disease/phenotype entries that contain an INCLUDED entry since this also happens with https://omim.org/entry/605074 and the omim.ttl file only contains 1 disease to gene association ('has material basis in germline mutation in' PRCC). 

I believe this is part of the issues that were reported in the [PR ](https://github.com/monarch-initiative/mondo/pull/8108#issuecomment-2322123779) for the OMIM g2d pipeline in Mondo, specifically point (3) Genes should not be added if the OMIM record is associated with multiple genes.

UPDATE: 
For https://omim.org/entry/613659, I looked in the ttl file directly vs. from Protege and do see 11 entries in this format where the values like `_:Nbcf6a815046747ee9fe5bc8f3891b1c5` look to point back to a gene:
```
_:Nbcf6a815046747ee9fe5bc8f3891b1c5 a owl:Restriction ;
    owl:onProperty RO:0004013 ; 
    owl:someValuesFrom OMIM:613659 .
```
In 10 of the 11 entries, it has `owl:onProperty RO:0003302 ;`. None have RO:0004003 as displayed in Protege. 

UPDATE 2: I do see that the OMIM gene entry with RO:0004013 is for IL1B and there is some code that flips RO:0004013 to RO:0004003 so those further transformations are more clear. 

**LATEST QUESTION** --> However, it's not clear why only this one gene has RO:0004013 to start with and the others listed for 'gastric cancer' have a different RO property.

------
Also, did anyone have a chance to document the earlier design decisions? See https://github.com/monarch-initiative/omim/issues/75#issuecomment-1320976773

## Resources
- [Google Sheet](https://docs.google.com/spreadsheets/d/1-9V5nnsCc1rHXzzVrr9OuE9bTHTGYptT4QWP5kiTYRM/edit?gid=1792806750#gid=1792806750) of before/after results from #158
- [Tracker / discussion doc](https://docs.google.com/document/d/1cLfBgPIZWiN5LX-E-xwSyBeFdT-vw0JuSfSM7HL3_hc/edit?tab=t.0)

### PR DESCRIPTION
**Pull request by joeflack4**
resolves #156

## Changes
- Bug fix: Code for adding Disease2Gene relationships. These should be added if and only if there is one gene associated with the disease, unless it is a digenic disease/phenotype. However before, we were erroneously checking the number of diseases associated with a given gene.
- Delete: Code for adding Gene2Disease relationships
- Add; SPARQL query and goal to list relationships
- Added and updated comments related to gene disease associations

## Resources
- [Google Sheet](https://docs.google.com/spreadsheets/d/1-9V5nnsCc1rHXzzVrr9OuE9bTHTGYptT4QWP5kiTYRM/edit?gid=25340125#gid=25340125) of before/after results
- [Tracker / discussion doc](https://docs.google.com/document/d/1cLfBgPIZWiN5LX-E-xwSyBeFdT-vw0JuSfSM7HL3_hc/edit?tab=t.0)

### CODE REVIEW

File: omim2obo/main.py [hunk 1]
```diff
@@ -280,49 +280,61 @@ def omim2obo(use_cache: bool = False):
...
+
+    # - Cyto location: Add RO:0002525 (is subsequence of)
+    # https://www.ebi.ac.uk/ols/ontologies/ro/properties?iri=http://purl.obolibrary.org/obo/RO_0002525
+    for gene_mim, gene_data in gene_phenotypes.items():
+        if gene_data['cyto_location']:
+            chr_id = '9606chr' + gene_data['cyto_location']  # todo: document meaning of 9606chr
+            add_subclassof_restriction(graph, RO['0002525'], CHR[chr_id], OMIM[gene_mim])
+
+    # TODO: @Trish: I've removed addition of gene->disease relations, leaving only disease->gene. Is that what we want?
+    #  Removed because extra complexity. Want to get down disease->gene code first. And Mondo doesn't need them.
```
Comment (twhetzel) [inline omim2obo/main.py:293]: We discussed calculating the number of associations [here](https://github.com/monarch-initiative/omim/pull/158/files#diff-6a2c7728d655af0d58b05a0d99cc2865c94c355a91d8df1d647a002f7eee6a4eL307) differently since it did not appear that if a gene was unique in the morbidmap file that it was sufficient to tag that gene as causal for a disease, and instead that the total number of genes per phenotype mim number should be used to determine if a gene is causal (ie only 1 gene per disease).

Comment (joeflack4) [reply omim2obo/main.py:293]: 1. Gene removal
There are a few reasons. It's been a year since this was written, so I don't remember everything, but I keep being reminded about the disease-centric nature of things, and I wonder whether we meant to add them in the first place. And moreso, I want this code review to be easier at first; for us to get the disease-gene stuff handled first. If you look at the code, I had to add on a bit to what you wrote to complete it.
Thinking more about it today though, I did leave part of the gene restrictions; the "cyto location" bit. And for adding the other ones back, I think the best way to handle it is to add some extra lines to this "cyto location" block. It'll actually look very similar to what was there before; I just need to drop the "n associations" conditional for the genes as that only matters for the diseases.
TLDR: Let's make sure the Disease2Gene stuff is correct first and I can add back the Gene2Disease shortly.

2. Yes, AFAIK only `mondo-ingest` consumes these outputs.

File: omim2obo/main.py [hunk 2]
```diff
@@ -280,49 +280,61 @@ def omim2obo(use_cache: bool = False):
...
-            add_subclassof_restriction_with_evidence(graph, predicate, OMIM[p_mim], OMIM[mim_number], evidence)
-            if predicate in MORBIDMAP_PHENOTYPE_MAPPING_KEY_INVERSE_PREDICATES:
-                # The following code basically just adds the converse relation:
-                # If there is g2d in the previous code, we now, in addition, add d2g
-                inverse_predicate = MORBIDMAP_PHENOTYPE_MAPPING_KEY_INVERSE_PREDICATES[predicate]
-                add_subclassof_restriction_with_evidence(
-                    graph, inverse_predicate, OMIM[mim_number], OMIM[p_mim], evidence)
+            gene_mim, p_lab, p_map_key, p_map_lab = assoc['gene_id'], assoc['phenotype_label'], \
+                assoc['mapping_key'], assoc['mapping_label']
+            # TODO: @Trish: Decide: Should have more relationships now because I simplified to what Sabrina asked.
```
Comment (joeflack4) [inline omim2obo/main.py:325]: ## Marking casual (`RO:0004003`)
I'm not sure if this is what we want. Please review.
- a. Sabrina said that if 1 gene assoc w/ disease, it's causal, and we can declare it "RO:0004003 (has material basis in germline mutation in)".
- b. But before, we were only doing that if OMIM specified the gene->disease relation as "'3': 'The molecular basis for the disorder is known; a mutation has been found in the gene.',", which we have associated with RO:0004013 (is causal germline mutation in), and that is the inverse of RO:0004003

Comment (twhetzel) [reply omim2obo/main.py:325]: I think the previous check can remain in place. Confirm that only 1 gene unless digenic and also make sure it has mapping key (3).

Generally I think that any disease that has only gene listed listed in the phenotype-gene table for the entry (or in morbidmap, maybe found in the other files too), will all have mapping key (3) based on they key definition "The molecular basis for the disorder is known; a mutation has been found in the gene.". However, there were entries in the Excel file that had two lines in morbidmap and the mapping keys were different between these lines in morbidmap.

Comment (twhetzel) [reply omim2obo/main.py:325]: @joeflack4 I have analyzed the morbidmap file and there are either a combination of issues with data in that file and/or other edge cases we need to consider to determine if a gene is causal.

Comment (joeflack4) [reply omim2obo/main.py:325]: Decided that yes, we'll only do mapping key = 3 again. Resolving.

Comment (twhetzel) [reply omim2obo/main.py:325]: Also, note my comment about digenic diseases is also no longer valid based on Sabrina's last email (Gene association in Mondo) on the rule or did you interpret her comment that the handful of diseases that have digenic in the phenotype label and also have only 1 gene association and have mapping key 3 should be included?

File: omim2obo/main.py [hunk 3]
```diff
@@ -280,49 +279,45 @@ def omim2obo(use_cache: bool = False):
...
+            # if p_map_key == '1':  # association w/ unknown defect; skip
+            #     continue  # See: https://github.com/monarch-initiative/omim/issues/79#issuecomment-1319408780)
+            phenotype_genes[p_mim].append({
+                'gene_id': gene_mim, 'phenotype_label': p_lab, 'mapping_key': p_map_key, 'mapping_label': p_map_lab})
+
+    # - Disease-Gene relations, 2 of 2: Add relations (subclass restrictions)
+    #   Only add if (1) only 1 association (2)‚mapping key = 3, (3) 'definitive'
+    phenotype_genes2: Dict[str, Dict[str, str]] = {k: v[0] for k, v in phenotype_genes.items() if len(v) == 1}  # 1assoc
+    phenotype_genes2 = {k: v for k, v in phenotype_genes2.items() if p2g_is_definitive(v['phenotype_label'])}
+    for p_mim, assoc in phenotype_genes2.items():
```
Comment (twhetzel) [inline omim2obo/main.py:312]: I believe this will still include the handful of diseases where [p_label contains the word 'digenic'](https://docs.google.com/document/d/1cLfBgPIZWiN5LX-E-xwSyBeFdT-vw0JuSfSM7HL3_hc/edit?tab=t.0#heading=h.h4y343h64cck) since there are cases where these exist with 1 gene association and a mapping key of 3 and are not prefixed with an omim special character.

Comment (joeflack4) [reply omim2obo/main.py:312]: Yes, that's true. I was going to follow up with you about that.

Basically my rationale for not filtering these out is because of the OMIM team's response. It seems to me that they haven't really insured the data quality of digenic diseases. So I suppose that could mean one of the following or both:
1. Non digenic, but being labeled as such
2. Digenic but they only labeled one of the associations

I had the first case at the forefront of my mind. But now that I think about it, for these cases where it is labeled diegenic but it only has one association, we have no idea which of these two cases it is. 

So therefore, I do agree it is a good idea to filter them out now. 

However I wonder if we want to consider outputting them into a special table for curator review. Perhaps some of them are not actually digenic?

Comment (twhetzel) [reply omim2obo/main.py:312]: Sure, I think outputting them to a file for curator review is a good idea!

File: omim2obo/main.py (hunk 1 above)
Comment (joeflack4) [reply omim2obo/main.py:292]: Added the gene disease relationships back. 

I had to put some thought into it to make sure I felt it was correct, and to make the code flow as cleanly and understandably as I could make it. 

Feel free to review here or during meeting.

File: omim2obo/main.py (hunk 3 above)
Comment (joeflack4) [reply omim2obo/main.py]: Done!
I added `review.tsv` to the release files and documentation for it, including the types of cases found inside of it, in the `README.md`.

- [`review.tsv` Google Sheet](https://docs.google.com/spreadsheets/d/1ftVMjsAtzVLv4aZFQ7KelRfFjct4VEprn6E76EdDJgw/edit?gid=441961077#gid=441961077)

File: omim2obo/main.py [hunk 4]
```diff
@@ -280,49 +280,72 @@ def omim2obo(use_cache: bool = False):
...
-        assocs: List[Dict] = mim_data['phenotype_associations']
-        # - add associations
+    gene_phenotypes: Dict[str, Dict] = parse_morbid_map(get_mim_file('morbidmap', download_files_tf))
+
+    # Gene-Chromosome relationships
+    # - Cyto location: Add RO:0002525 (is subsequence of)
+    # https://www.ebi.ac.uk/ols/ontologies/ro/properties?iri=http://purl.obolibrary.org/obo/RO_0002525
+    for gene_mim, gene_data in gene_phenotypes.items():
+        if gene_data['cyto_location']:
+            chr_id = '9606chr' + gene_data['cyto_location']  # todo: document meaning of 9606chr
```
Comment (twhetzel) [inline omim2obo/main.py:290]: 9606 is the ncbi taxonomy identifier for Homo sapiens

Comment (joeflack4) [reply omim2obo/main.py:290]: thank you!
i'll add that

**Comment by joeflack4** [comment]
@twhetzel This is no problem! It's on my schedule to do before Monday.
